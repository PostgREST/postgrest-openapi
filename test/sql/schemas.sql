-- detects tables as objects
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'type');

-- detects columns with enum types as an enum property
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'properties'->'size');

-- references a composite type column from a table to a component definition
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'properties'->'attr');

-- identifies the required columns
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'required');

-- detects comments done on a table and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'description');

-- detects comments done on a table column and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'properties'->'id'->'description');

-- maps sql types to OpenAPI types correctly
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'openapi_types');

-- does not show tables outside the exposed schema
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'secret_table');

-- detects composite types as objects
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'type');

-- references a composite type column from another composite type to a component definition
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'properties'->'dim');

-- detects a composite type inside another composite type
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.dimension');

-- detects an array composite type inside another composite type
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.color');

-- does not show a composite type that is not used by an exposed table
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.hiddentype');

-- detects comments done on a composite type and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'description');

-- detects comments done on a composite type column and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'properties'->'other'->'description');

-- references a composite type column from an array composite type inside the items property
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'properties'->'colors'->'items');

-- defines all the available prefer headers
select key, jsonb_pretty(value)
from jsonb_each(get_postgrest_openapi_spec('{test}')->'components'->'schemas')
where key like 'header.prefer%';
