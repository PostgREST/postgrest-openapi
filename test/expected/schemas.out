-- detects tables as objects
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'type');
 jsonb_pretty 
--------------
 "object"
(1 row)

-- detects columns with enum types as an enum property
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'properties'->'size');
        jsonb_pretty        
----------------------------
 {                         +
     "enum": [             +
         "XS",             +
         "S",              +
         "M",              +
         "L",              +
         "XL"              +
     ],                    +
     "type": "string",     +
     "format": "types.size"+
 }
(1 row)

-- references a composite type column from a table to a component definition
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'properties'->'attr');
                    jsonb_pretty                    
----------------------------------------------------
 {                                                 +
     "$ref": "#/components/schemas/types.attribute"+
 }
(1 row)

-- identifies the required columns
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'required');
   jsonb_pretty    
-------------------
 [                +
     "id",        +
     "code",      +
     "name",      +
     "description"+
 ]
(1 row)

-- detects comments done on a table and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'description');
                              jsonb_pretty                               
-------------------------------------------------------------------------
 "Products summary\n\nProducts description\nthat spans\nmultiple lines."
(1 row)

-- detects comments done on a table column and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'products'->'properties'->'id'->'description');
       jsonb_pretty        
---------------------------
 "identifier of a product"
(1 row)

-- maps sql types to OpenAPI types correctly
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'openapi_types');
                 jsonb_pretty                  
-----------------------------------------------
 {                                            +
     "type": "object",                        +
     "properties": {                          +
         "a_json": {                          +
             "type": "object",                +
             "format": "json"                 +
         },                                   +
         "a_real": {                          +
             "type": "number",                +
             "format": "real"                 +
         },                                   +
         "a_text": {                          +
             "type": "string",                +
             "format": "text"                 +
         },                                   +
         "a_jsonb": {                         +
             "type": "object",                +
             "format": "jsonb"                +
         },                                   +
         "a_bigint": {                        +
             "type": "integer",               +
             "format": "bigint"               +
         },                                   +
         "a_boolean": {                       +
             "type": "boolean",               +
             "format": "boolean"              +
         },                                   +
         "a_int_arr": {                       +
             "type": "array",                 +
             "items": {                       +
                 "type": "integer",           +
                 "format": "integer"          +
             },                               +
             "format": "integer[]"            +
         },                                   +
         "a_integer": {                       +
             "type": "integer",               +
             "format": "integer"              +
         },                                   +
         "a_numeric": {                       +
             "type": "number",                +
             "format": "numeric"              +
         },                                   +
         "a_bool_arr": {                      +
             "type": "array",                 +
             "items": {                       +
                 "type": "boolean",           +
                 "format": "boolean"          +
             },                               +
             "format": "boolean[]"            +
         },                                   +
         "a_char_arr": {                      +
             "type": "array",                 +
             "items": {                       +
                 "type": "string",            +
                 "format": "character"        +
             },                               +
             "format": "character[]"          +
         },                                   +
         "a_json_arr": {                      +
             "type": "array",                 +
             "items": {                       +
                 "type": "object",            +
                 "format": "json"             +
             },                               +
             "format": "json[]"               +
         },                                   +
         "a_smallint": {                      +
             "type": "integer",               +
             "format": "smallint"             +
         },                                   +
         "a_text_arr": {                      +
             "type": "array",                 +
             "items": {                       +
                 "type": "string",            +
                 "format": "text"             +
             },                               +
             "format": "text[]"               +
         },                                   +
         "a_character": {                     +
             "type": "string",                +
             "format": "character",           +
             "maxLength": 1                   +
         },                                   +
         "a_jsonb_arr": {                     +
             "type": "array",                 +
             "items": {                       +
                 "type": "object",            +
                 "format": "jsonb"            +
             },                               +
             "format": "jsonb[]"              +
         },                                   +
         "a_bigint_arr": {                    +
             "type": "array",                 +
             "items": {                       +
                 "type": "integer",           +
                 "format": "bigint"           +
             },                               +
             "format": "bigint[]"             +
         },                                   +
         "a_numeric_arr": {                   +
             "type": "array",                 +
             "items": {                       +
                 "type": "number",            +
                 "format": "numeric"          +
             },                               +
             "format": "numeric[]"            +
         },                                   +
         "a_varchar_arr": {                   +
             "type": "array",                 +
             "items": {                       +
                 "type": "string",            +
                 "format": "character varying"+
             },                               +
             "format": "character varying[]"  +
         },                                   +
         "a_double_precision": {              +
             "type": "number",                +
             "format": "double precision"     +
         },                                   +
         "a_character_varying": {             +
             "type": "string",                +
             "format": "character varying"    +
         }                                    +
     }                                        +
 }
(1 row)

-- does not show tables outside the exposed schema
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'secret_table');
 jsonb_pretty 
--------------
 
(1 row)

-- detects composite types as objects
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'type');
 jsonb_pretty 
--------------
 "object"
(1 row)

-- references a composite type column from another composite type to a component definition
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'properties'->'dim');
                    jsonb_pretty                    
----------------------------------------------------
 {                                                 +
     "$ref": "#/components/schemas/types.dimension"+
 }
(1 row)

-- detects a composite type inside another composite type
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.dimension');
          jsonb_pretty           
---------------------------------
 {                              +
     "type": "object",          +
     "properties": {            +
         "hei": {               +
             "type": "number",  +
             "format": "numeric"+
         },                     +
         "len": {               +
             "type": "number",  +
             "format": "numeric"+
         },                     +
         "wid": {               +
             "type": "number",  +
             "format": "numeric"+
         }                      +
     }                          +
 }
(1 row)

-- detects an array composite type inside another composite type
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.color');
            jsonb_pretty            
------------------------------------
 {                                 +
     "type": "object",             +
     "properties": {               +
         "def": {                  +
             "type": "string",     +
             "format": "text"      +
         },                        +
         "hex": {                  +
             "type": "string",     +
             "format": "character",+
             "maxLength": 6        +
         }                         +
     }                             +
 }
(1 row)

-- does not show a composite type that is not used by an exposed table
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.hiddentype');
 jsonb_pretty 
--------------
 
(1 row)

-- detects comments done on a composite type and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'description');
                               jsonb_pretty                                
---------------------------------------------------------------------------
 "Attribute summary\n\nAttribute description\nthat spans\nmultiple lines."
(1 row)

-- detects comments done on a composite type column and shows it as description
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'properties'->'other'->'description');
              jsonb_pretty               
-----------------------------------------
 "other information about the attribute"
(1 row)

-- references a composite type column from an array composite type inside the items property
select jsonb_pretty(get_postgrest_openapi_spec('{test}')->'components'->'schemas'->'types.attribute'->'properties'->'colors'->'items');
                  jsonb_pretty                  
------------------------------------------------
 {                                             +
     "$ref": "#/components/schemas/types.color"+
 }
(1 row)

-- defines all the available prefer headers
select key, jsonb_pretty(value)
from jsonb_each(get_postgrest_openapi_spec('{public}')->'components'->'schemas')
where key like 'header.prefer%';
           key            |                               jsonb_pretty                               
--------------------------+--------------------------------------------------------------------------
 header.preferTx          | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "tx": {                                                         +
                          |             "enum": [                                                   +
                          |                 "commit",                                               +
                          |                 "rollback"                                              +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "description": "Specify how to end a transaction"           +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferCount       | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "count": {                                                      +
                          |             "enum": [                                                   +
                          |                 "exact",                                                +
                          |                 "planned",                                              +
                          |                 "estimated"                                             +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "description": "Get the total size of the table"            +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferParams      | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "params": {                                                     +
                          |             "enum": [                                                   +
                          |                 "single-object"                                         +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "deprecated": true,                                         +
                          |             "description": "Send JSON as a single parameter"            +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferReturn      | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "return": {                                                     +
                          |             "enum": [                                                   +
                          |                 "minimal",                                              +
                          |                 "headers-only",                                         +
                          |                 "representation"                                        +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "default": "minimal",                                       +
                          |             "description": "Return information of the affected resource"+
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferMissing     | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "missing": {                                                    +
                          |             "enum": [                                                   +
                          |                 "default",                                              +
                          |                 "null"                                                  +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "default": "null",                                          +
                          |             "description": "Handle null values in bulk inserts"         +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferHandling    | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "handling": {                                                   +
                          |             "enum": [                                                   +
                          |                 "strict",                                               +
                          |                 "lenient"                                               +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "default": "lenient",                                       +
                          |             "description": "How to handle invalid preferences"          +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferTimezone    | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "timezone": {                                                   +
                          |             "type": "string",                                           +
                          |             "description": "Specify the time zone"                      +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferResolution  | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "resolution": {                                                 +
                          |             "enum": [                                                   +
                          |                 "merge-duplicates",                                     +
                          |                 "ignore-duplicates"                                     +
                          |             ],                                                          +
                          |             "type": "string",                                           +
                          |             "description": "Handle duplicates in an upsert"             +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
 header.preferMaxAffected | {                                                                       +
                          |     "type": "object",                                                   +
                          |     "properties": {                                                     +
                          |         "max-affected": {                                               +
                          |             "type": "integer",                                          +
                          |             "description": "Specify the amount of resources affected"   +
                          |         }                                                               +
                          |     }                                                                   +
                          | }
(9 rows)

